toc/rfc/rfc-draft-ssh-over-ws.md

```diff
@@ -0,0 +1,329 @@
+
+These scans usually complain that besides port 443 for HTTP traffic, also port 2222 is open for SSH on the internet-facing load balancer. Many customers demand closure of the port as they
+view it as a potential attack surface for brute-force attacks against the SSH protocol.
+
+The issue is exacerbated by features like "bring-your-own-domain" where customers provide a CNAME entry alongside a matching x509 certificate that points their own (sub-)domain to an app hosted on the shared Cloud Foundry environment.
+
+This can cause potential reputation damage to the customer because to an outsider, it looks like the port 2222 is open on the customer's domain, e.g. `www.big-corp.com:2222` open looks much worse than `my-app.cfapps.eu10.hana.ondemand.com:2222` even though both are technically the same.
+
+Security-wise it is easier for an attacker to probe the environment by simply doing a port scan and then testing direct attacks on the SSH proxy:
+
+```
+nmap api.cf.eu10.hana.ondemand.com
+
+Starting Nmap 7.94 ( https://nmap.org ) at 2024-04-08 10:29 CEST
+Nmap scan report for api.cf.eu10.hana.ondemand.com (3.124.222.77)
+Host is up (0.031s latency).
+Other addresses for api.cf.eu10.hana.ondemand.com (not scanned): 3.122.209.241 3.124.208.223
+rDNS record for 3.124.222.77: ec2-3-124-222-77.eu-central-1.compute.amazonaws.com
+Not shown: 997 filtered tcp ports (no-response)
+PORT     STATE SERVICE
+80/tcp   open  http
+443/tcp  open  https
+2222/tcp open  EtherNetIP-1
+```
+
+then proceed with connecting to the port 2222:
+
+```
+nc api.cf.eu10.hana.ondemand.com 2222
+
+SSH-2.0-diego-ssh-proxy
+```
+
+the SSH proxy greets with the SSH-2.0 version, allowing the attacker to try all kinds of SSH-2.0 related probes.
+
+Another facet of the problem is that many customers require explicit allow-listing of external web sites from their on-prem environments.
+Port 443 is usually not a problem as many firewalls open it by default, but port 2222 likely is frowned upon and cause for concern.
+
+## Proposal
+
+### Note on Security
+The proposal does not aim to solve security issues of (CF) SSH per se. Tunneling a protocol through another does not make it inherently more secure. Any weakness of SSH within the CF SSH proxy will still be present even if the RFC is accepted. The main goal is to reduce the number of open ports required. Although, there is one minor security improvement as the SSH traffic will then be also encrypted by TLS within the WebSocket tunnel.
+
+
+### Architecture Changes
+
+![architecture-changes](rfc-draft-ssh-over-ws/cf-ssh-over-websocket.drawio.png)
+
+There are three scenarios:
+
+#### Scenario A: Legacy SSH
+This is the default scenario as of today. All changed components MUST ensure this scenario still works. It SHOULD be tested with legacy CLI versions such as v6 which are no longer updated. This scenario includes regular SSH clients like OpenSSH.
+
+#### Scenario B: Dual SSH / WS
```

Comment:
Which scenarios do you have in mind to also support the mixed mode?
I currently do not see the benefit for an operator; if the port cannot be closed, are the additional switches worth the gain of partial websocket usage?

Reply:
buying time for legacy clients basically. you want to migrate customers off the 2222 port while also allowing frontrunners to use the new feature. the mixed mode allows you to do a gradual shift instead of a big bang.
if you can't run mixed you won't have any time to collect feedback ahead of time and cannot easily roll back in case of any show stoppers that might turn up.